### ISSUE TITLE
http: websockets should interact with full http filter chain

### ISSUE THREAD

**Issue by mikedanese** [envoyproxy/envoy#3301]
#### Example Problem

Currently, the http connection manager immediately surrenders websocket requests to the websocket handler, bypassing all http_filters. This is problematic for API front proxies which need to terminate client TLS and pass authentication info in headers to upstream API servers. An Http::StreamDecoderFilter implementing such a proxy will not be called in the websocket handling code path, effectively disallowing TLS client auth on requests requiring websockets.

**Comment by mikedanese** [comment]
The websocket spec [requires](https://tools.ietf.org/html/rfc6455#section-1.2) that the server responds with upgrade headers checked in [isWebsocketUpgradeRequesrt](https://github.com/envoyproxy/envoy/blob/833769c93ccaae9c7c91d7b01970a750428b5ab9/source/common/http/utility.h#L91):

```
Upgrade: websocket
Connection: Upgrade
```

A fairly general solution to this problem would be to defer handing off the connection to the websocket manager until after a successful response is received.

This would also allow recycling of http connections on failed upgrade requests to websocket handlers (e.g. when the upstream returns a 403 on a websocket upgrade request). Today, these failed connections need to be explicitly closed by the upstream.

**Comment by mikedanese** [comment]
cc @rshriram @mattklein123

**Comment by mattklein123** [comment]
@mikedanese the functionality change definitely makes sense to me. I think there are a few options here on how we could fix things. Unfortunately none of them are going to be easy.
1. Modify as you suggest so that we end up proxying the initial request all the way through to the router and back, and then do the hand off. There are some gotchas here that aren't immediately clear to me how to solve.
a) Early data (this was fixed recently). I think the data would have to be held by the connection manager until the response is received.
b) When the connection manager switches over to WS, the existing filters likely need to be destroyed. I generally have concerns about filters not being WS aware, and they may basically have to be reset to work properly (and then will that yield messed up stats, etc.). This in general makes me wonder if we need HTTP filters to actually be WS aware in some way.

2. Don't actually fully proxy the WS request to upstream, but allow filter iteration from the connection manager for WS requests. This would have to respect filter stop, deal with early data, and have all the same problems about filters being reset as (1), so it's probably about the same amount of work.

I will admit I don't have a very clear proposal right now without sitting down and doing some more thinking, so hopefully someone else can dig in and come up with a formal design proposal. @rshriram @ggreenway @alyssawilk for more comments.

**Comment by mattklein123** [comment]
> Matt, for b) were you thinking of handling the WS upgrade as a header only request/response, then handing off any data and the rest of the connection to the tcp proxy session as a "second request"? I think that could be done pretty cleanly but you're more familiar with the data plane than I am.

Yeah, I think this is exactly how it would have to work, and I think would be the cleanest solution. We can dispatch a "request" and then depending on the response, decide what to do from there inside the connection manager.

**Comment by rshriram** [comment]
One alternative is to add a new codec that supports the bare minimal websocket stuff framing. While http filters can be used with websockets, filters would typically see only the first req/resp and then the rest is opaque or even simply an extended decodeData()?

I haven’t revisited this in a while. But this was one initial line of thought.

**Comment by alyssawilk** [comment]
It'd be helpful to us if websocket ugrades traversing an H2 hop could have full header sanitation on the H2 hops and onData calls for the websocket payload (for say DoS byte accounting).  I haven't looked at the websocket spec lately, but if we just allow pass-through of the upgrade header and make sure that bidirectional H2 works correctly, would this "Just Work"?  Anyone object to allowing this as opposed to only supporting CONNECT style upgrades as suggested on #1630?

**Comment by alyssawilk** [comment]
Ok, fair, maybe I do want #1630.  I've filed https://github.com/nghttp2/nghttp2/issues/1181 to see if we can get (or add) nghttp2 support for this.

That would leave us with H2 "websockets" being treated as HTTP for both header and data accounting and H1 being treated as HTTP for headers and TCP proxying for payload.  Per 1) b) *if* I can sort out any bidi-HTTP/1 issues, would anyone object to at least having a configurable option for the HTTP/1.1 websockets for passing not just headers but also body through the HTTP filter chain rather than doing TCP proxy handoff?    As Matt points out the filters need to be websocket-aware in any case, and I can't speak for others but given the current things we do onData (mostly byte counting to various data sinks) it's easier for us to apply existing HTTP filters to websocket payload than detach them and implement the same onData functions on the TCP side and it's nice to have things accounted for consistently.

**Comment by alyssawilk** [comment]
Also @ggreenway as he's been doing websocket changes and may have an opinion about adding more complex paths :-P

**Comment by ggreenway** [comment]
WebSockets is such an awkward thing to handle here. If we pass it through most of the http stack, you have to make all of those bits of the stack handle some non-http. If not all the filters support this, you'll have to carefully document which filters can be used with WebSockets. And we'll be adding complexity to all the filters that do handle WebSockets. If we don't go through the http code, we lose many of the benefits of WebSockets starting out as http.

I think I'd be more in favor of only handling the part of the data that really is http in the http-filters, to avoid complexity leaking into every filter. And then figure out some way to share the code, or at least make common the stats/accounting, for the bidirectional websocket data after the initial http handshake is complete.

**Comment by mattklein123** [comment]
@ggreenway that was my initial suggestion. I think if we do this the filters only see a header only request for the upgrade. That's all.

**Comment by ggreenway** [comment]
That seems reasonable. I don't think the code for http/1 WebSockets will be too bad. I think http/2 will be stranger because we can't drop to a TcpProxy; we'll still have to go through at the bottom part of the h2 stack.

**Comment by alyssawilk** [comment]
Yeah, that's precisely my concern.   I'd started playing around with #1630 and I think we need some discussion of how we want to handle websockets in general in Envoy since I'd prefer we handle both paths consistently.

My assumption is if we went with an H2-H2 connect hop or H1-H2 websocket->CONNECT upgrade per #1630, the entire websocket payload would pass through the filters because that path is the only way we can send the data out through a single outbound H2 stream.   If we went that route, each filter is going to have to handle websocket (and other CONNECT) payload anyway.  I'd rather consistently pass things through onData for the current H1-H1 and the H2-H2 CONNECT->websocket downgrade (also per 1630) than have things handled significantly differently for H2 upstream and H1 upstream with the TCP proxy hand-off for one case and treating it like HTTP for the other.   

The alternate is that we come up with some clever idea of how to jam a TCP proxy filter stack in the H2 path bridging the HttpConnectionManager directly to the H2 stream the router is writing to.  Beyond not seeing a really clean way to do that, the TCP stack seems the wrong abstraction for this case.  It's really not a standalone connection so for example doing idle timeouts if the stream stalls due to being preempted by higher priority streams (once we handle priorities) won't be correct.  

For another way of looking at it, as TCP filters should support on[Raw]Data, HTTP filters should support onHeaders/on[HTTP]Data, and ideal websocket/connect filters should support onHeaders, on[Raw]Data.  Right now we're bridging the two by trying to pass the headers through HTTP filters and data through TCP filters, but the data will have to pass through an HTTP onData for serialization H2 path regardless.  Given this, maybe there's some clean way of making it clear if any HTTP filter is websocket-friendly and the HCM can just bypass any filter which should not apply?

Happy to schedule a quick chat over hangouts/zoom or we can discuss this on Tuesday if either would be easier.

**Comment by ggreenway** [comment]
Yeah, let's schedule some time to discuss this and figure out the best path forward.

**Comment by mattklein123** [comment]
I think the best thing to do here might be to have a separate filter stack for web socket connections. The filter stack construction code would have to be changed slightly, but I don't think it would be very difficult to wait until initial headers are received. WS filters can be different types, and optionally a filter could implement both types so it could be used in both filter stacks. This would be a medium/large project, but would probably leave us in the cleanest place afterwards. 

But yes we should definitely discuss live to brainstorm what to do.

**Comment by alyssawilk** [comment]
The summary of the conversation of today's community meeting is we're going to in configuration allow both an HTTP filter chain and a websocket filter chain, where for now in Envoy we'll reuse existing HTTP Filter APIs so folks can reuse their filters for both if they so choose.  

Plan A involves using the existing router filter for websocket-to-upstream, with the caveat if we find that upstream connections don't handle pipelined HTTP followed by a websocket upgrade, there may be follow-up work to ensure we use a fresh connection for the upgrade rather than a recycled connection.

The other thing discussed was how to handle the H1/H2 upgrades/downgrades.  Envoy currently consistently normalizes on HTTP/2 semantics, so that filters don't have to branch where H1 and H2 differ.  In this one case we definitely want to normalize (imagine someone wanting to have a CORS filter blocking based on method, where it would block differently depending on if it was H1 websocket (GET with upgrade header) or H2 websocket (CONNECT style).  I'm mildly inclined to in this one case to normalize on HTTP/1.1 style (downgrade incoming H2 CONNECT to GET-with-upgrade before creating the filters) since websocket really is more of an HTTP/1.1 concept, but if anyone has even a mild preference for consistently treating things as H2 I'm just as happy to go that route.  Speak now!   

The current websocket implementation (header intercept and TCP proxy handoff) will be supported until the new version is implemented, and then will go through the usual Envoy deprecation cycle to give folks plenty of time to both test and deploy the new way of doing things.

**Comment by alyssawilk** [comment]
Hey @mattklein123 and @ggreenway 

My proposed API changes for "new style" websockets basically allow a custom filter chain for any upgrade type configured on a per-listener basis (adding it to the HttpConnectionManager for now).  My thought was that'd be a bit more flexible for other upgrades folks might want as well as being reusable for connect support.  SG?

  // The configuration for HTTP upgrades.
  // For each upgrade type desired, an UpgradeConfig must be added.
  message UpgradeConfig {
    // The case-insensitive name of this upgrade, e.g. "websocket".
    // For each upgrade type present in upgrade_configs, requests with
    // Upgrade: [upgrade_type]
    // will be proxied upstream.
    string upgrade_type = 1;
    // If present, this represents the filter chain which will be created for
    // this type of upgrade.  If no filters are present, the filter chain for
    // HTTP connections will be used for this upgrade type.
    repeated HttpFilter filters = 5;
  };
  repeated UpgradeConfig upgrade_configs = 23;

**Comment by rshriram** [comment]
To confirm this is going inside http connection manager ? Is there a possibility to add network filters? For example, you could imagine streaming arbitrary tcp connections (eg Mongo or MySQL over a h2 stream just like websockets. And these things would benefit from a network filter more than http filter).

**Comment by alyssawilk** [comment]
Neither the current implementation or the changes discussed here allows attaching a network filter.  We're basically just reusing the HTTP is processing [HTTP-headers, HTTP-data] for processing [HTTP-headers, Arbitrary-Data].  I don't think it'd be a bad thing to support network filters for the data processing part but the problem is if we use a HTTP/TCP hybrid, we end up with a big disparity between how we handle upgrades/connect for H1 and H2 (where we simply can't TCP proxy since it's still stream based, so it basically has to go through something akin to the HTTP filter APIs at some point)

It's not the most elegant workaround ever, but in the short run one could write an http filter adapter/wrapper for any network filters to simply pass the "HTTP" onData to the network filter's onData functions.   Ideally in the long run we'd actually change the UpgradeConfig to take a WebSocket filter which could basically wrapper HTTP filters or TCP proxy filters, but I think I'm only doing the first step to getting us there.

**Comment by mattklein123** [comment]
@alyssawilk LGTM

**Comment by ggreenway** [comment]
@alyssawilk What other upgrade types are possible? Do http filters make sense for them? We're kind of shoe-horning http filters in for websockets, because it's close enough. Are there any other potential upgrade types where that won't make sense?

**Comment by alyssawilk** [comment]
I was thinking connect upgrades might want to go this path, and while right now we're only parsing upgrade headers, I was thinking of adding a "magic keyword" for connect which was unlikely to overlap with anything real in HTTP land.

But really for all upgrades, there's the upgrade header path and then a bunch of onData calls, so I think the HTTP filters (which take headers and then data) are a more logical API match.

**Comment by ggreenway** [comment]
Sounds reasonable to me.

**Comment by alyssawilk** [comment]
ok, I have a new awful corner case to discuss, which is websockets with chunked bodies (request or response bodies)

I believe technically they're allowed.  They're nonsensical for H2 since you wouldn't be able to differentiate the body payload from the websocket payload.  But we'd want the http-body/websocket-body delineation to survive an H1-H2-H1 hop.  I think the way to do that is just proxy the raw chunks over he H2 link, and put in a substitute header x-envoy-upgrade-encoding-was-chunked for the H2 hop which gets turned back into "transfer-encoding: chunked" for the subsequent H1 hop.

So preferences on if we accept
 1) only upgrades with no bodies
 2) only upgrades only with explicit content lengths
 3) upgrades with any sort of bodies?  
I'd prefer 1 or 3 for consistency but I'm up for whatever.  If I hear nothing back I'm going for 3 since I'm halfway to implementing that in any case.

@mattklein123 @ggreenway @PiotrSikora

**Comment by PiotrSikora** [comment]
Is this valid, though? Once WebSocket handshake is completed, the connection turns into bidirectional end-to-end bytestream that's unrelated to HTTP, and should be opaque to the proxy, so I don't believe that `Transfer-Encoding: chunked` is valid here. Also, WebSocket protocol uses its own data framing, so it shouldn't care about chunking.

If you're seeing this in the wild, then I suspect that the intermediary advertises `Transfer-Encoding: chunked`, but switches to the bytestream tunnel anyway and doesn't actually produce valid chunks.

Also, I'd like to hear a strong argument as to why do we need to use `x-envoy-upgrade-encoding-was-chunked` and send `Transfer-Encoding: chunked` to the upstream, since it's a hop-by-hop header.

**Comment by alyssawilk** [comment]
Once the websocket handshake is complete, the payload is now websocket and not chunked HTTP.  As I read it, the HTTP request-response should totally allowed to be chunked as it's valid HTTP.  Something like [this](https://github.com/alyssawilk/envoy/blob/websocket_delay/test/integration/websocket_integration_test.cc#L525) where there's a chunked body, a terminal chunk, then websocket payload.

I don't think anyone should intentionally do this, but I'm concerned there'd be some proxy which took a content-length: 0 request and transformed it to a T-E: chunked\r\n\r\n0\r\n\r\n style body.  Or something which auto-chunked on the response path as response bodies are more likely.  So I'd prefer we either completely disallow bodies (content-length: 0 or bust) or handle both cases.

If we agree this is a corner case we ought to handle (T-E chunked is currently accepted and broken in our current websocket implementation), when the data traversed the H1->H2 hop, we'd have to preserve the boundary between HTTP body data and websocket payload.  The best* way to do that I can think of in H2  is to literally proxy the raw H2 payload (chunks included), and then re-append the T-E:chunked header when doing the H1 downgrade.  We have to communicate that the H1 payload is chunked over the H1-H2 hop and you're not allowe to use T-E chunked header, hence my proposed header shenanigans.

*We could also invent a new framing layer, or buffer the full request body and append a content length. But the former is obviously more complicated and the latter is less obviously super-complicated (which I can explain offline) so I'm pretending I didn't think of either.

**Comment by mattklein123** [comment]
@alyssawilk i don't have a strong opinion here. My general feeling is to support as few things as possible and then wait for someone to complain about other permutations.

**Comment by rshriram** [comment]
Do we have to drop bodies all together ? We could support upgrades with bodies within limits of what Envoy accepts today. No chunking business. And no business of half http payload and half websocket payload. I don’t even think it’s possible since the client has to wait for the server to respond with upgrade accepted and then client starts the framing by setting its sequence numbers etc. are you thinking of a scenario where the server sends websocket payload with upgrade confirmation ?

**Comment by PiotrSikora** [comment]
@alyssawilk I don't believe that HTTP request body before switching to WebSocket is a thing, and even if there are peers out there that emit stuff like that, then the behavior seems to be undefined, and I'd rather reject such requests, than support this weirdness and add to the problem (...unless we have a very good reason to do it).

[RFC6455, sec. 1.7](https://tools.ietf.org/html/rfc6455#section-1.7) says:

>   The WebSocket Protocol is an independent TCP-based protocol.  Its
>   only relationship to HTTP is that its handshake is interpreted by
>   HTTP servers as an Upgrade request.

i.e. WebSocket is not HTTP request-response, but a protocol that (ab)uses HTTP Upgrade for its handshake.

Granted, if the upgrade fails, then in theory request continues as a HTTP protocol, but I don't think that's supported by anything out there.

On top of that, `GET` with request body is undefined as well, since [RFC7231, sec. 4.3.1](https://tools.ietf.org/html/rfc7231#section-4.3.1) says:

>   A payload within a GET request message has no defined semantics;
>   sending a payload body on a GET request might cause some existing
>   implementations to reject the request.


FWIW, this is a tcpdump between Chrome & http://websocket.org/echo.html:

```
GET /?encoding=text HTTP/1.1
Host: echo.websocket.org
Connection: Upgrade
Pragma: no-cache
Cache-Control: no-cache
Upgrade: websocket
Origin: http://websocket.org
Sec-WebSocket-Version: 13
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.9,pl;q=0.8
Cookie: _ga=GA1.2.1696519655.1530136642; _gid=GA1.2.1985693603.1530136642; _gat=1; __zlcmid=n7hRtGQ5EUa36S
Sec-WebSocket-Key: JiUddN+YyhIb+TwvFCOFWA==
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
```

```
HTTP/1.1 101 Web Socket Protocol Handshake
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: content-type
Access-Control-Allow-Headers: authorization
Access-Control-Allow-Headers: x-websocket-extensions
Access-Control-Allow-Headers: x-websocket-version
Access-Control-Allow-Headers: x-websocket-protocol
Access-Control-Allow-Origin: http://websocket.org
Connection: Upgrade
Date: Wed, 27 Jun 2018 21:43:42 GMT
Sec-WebSocket-Accept: 2+qxxjue1vLopkSrGRTVv8nGZSk=
Server: Kaazing Gateway
Upgrade: websocket
```

No `Content-Length: 0` (since it's implied for `GET` requests) or `Transfer-Encoding` weirdness.

**Comment by alyssawilk** [comment]
Piotr, I'd be happy to block GET-with-body because I think it's an HTTP abomination, but where do you see that upgrades must be GETs?  From what I was reading they can be any valid HTTP request.

@rshriram again I'd really rather fully handle bodies or not handle bodies.  Right now the behavior allows T-E but is entirely broken and I don't like that state of affairs.   As said I also don't want to handle content-length style bodies and not TE as I've seen enough proxies which transform to be twitchy about handling one and not the other.  

Once I land my in-flight refactor I'll push out my WIP for handling both body types and folks can say yea or nay on the complexity with code to look at.  SG?

**Comment by ggreenway** [comment]
https://tools.ietf.org/html/rfc6455#section-4.1:

> The method of the request MUST be GET, and the HTTP version MUST be at least 1.1.

I agree with @PiotrSikora. Until we see a definite need, I'd say reject any body in an upgrade request.

**Comment by alyssawilk** [comment]
I'm happy to change the existing implementation and my outstanding new mode to reject both and document we do not currently support upgrades with bodies. Before I go ahead and do that, @rshriram as the one person who wants body support could you comment why?

**Comment by rshriram** [comment]
Not in request. I should have been clearer. Body support in the response. That said I just realized that we already support body in the response since the upstream connection is made as opaque tcp. Apologies for the confusion!

**Comment by alyssawilk** [comment]
Ah, that makes much more sense.  @PiotrSikora and @ggreenway thoughts on this?  The code for handling response bodies is basically the same as request bodies so if there's a legit request I'd be inclined to just handle bodies.

**Comment by MisterJimson** [comment]
What does opaque mean in this context?
